### PR TITLE
Add a middleware for authenticating users using Vkontakte social network

### PR DESCRIPTION
**Pull request by serber**
An ASP.NET Core middleware for authenticating users using Vkontakte (https://vk.com)

### CODE REVIEW

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationDefault.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿namespace AspNet.Security.OAuth.Vkontakte
+{
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationDefault.cs]: Please fix the bracing style to use K&R.

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationDefault.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    public static class VkontakteAuthenticationDefault
+    {
+        public const string AuthenticationScheme = "Vkontakte";
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationDefault.cs]: We need XML docs everywhere :clap:

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+{
+    internal class VkontakteAuthenticationHandler : OAuthHandler<VkontakteAuthenticationOptions>
+    {
+        public VkontakteAuthenticationHandler(HttpClient httpClient) : base(httpClient)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(ClaimsIdentity identity, AuthenticationProperties properties, OAuthTokenResponse tokens)
+        {
+            string endpoint = QueryHelpers.AddQueryString(Options.UserInformationEndpoint, "access_token", tokens.AccessToken);
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs]: `var`

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+            if (Options.Fields.Count > 0)
+            {
+                endpoint = QueryHelpers.AddQueryString(endpoint, "fields", string.Join(",", Options.Fields));
+            }
+
+            HttpResponseMessage response = await Backchannel.GetAsync(endpoint, Context.RequestAborted);
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs]: `var`

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            response.EnsureSuccessStatusCode();
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs]: Please replace that by https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.Gitter/GitterAuthenticationHandler.cs#L32-L39

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+            string responseText = await response.Content.ReadAsStringAsync();
+
+            JObject payload = JObject.Parse(responseText);
+            
+            AuthenticationTicket ticket = new AuthenticationTicket(new ClaimsPrincipal(identity), properties, Options.AuthenticationScheme);
+            OAuthCreatingTicketContext context = new OAuthCreatingTicketContext(ticket, Context, Options, Backchannel, tokens, payload);
+
+            string identifier = VkontakteAuthenticationHelper.GetId(payload);
+            if (!string.IsNullOrEmpty(identifier))
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs]: Use `AddOptionalClaim` everywhere.

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            response.EnsureSuccessStatusCode();
+
+            string responseText = await response.Content.ReadAsStringAsync();
+
+            JObject payload = JObject.Parse(responseText);
+            
+            AuthenticationTicket ticket = new AuthenticationTicket(new ClaimsPrincipal(identity), properties, Options.AuthenticationScheme);
+            OAuthCreatingTicketContext context = new OAuthCreatingTicketContext(ticket, Context, Options, Backchannel, tokens, payload);
+
+            string identifier = VkontakteAuthenticationHelper.GetId(payload);
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs]: `var` everywhere... :smile:

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationMiddleware.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        /// <param name="next">The next middleware in the HTTP pipeline to invoke.</param>
+        /// <param name="dataProtectionProvider"></param>
+        /// <param name="loggerFactory"></param>
+        /// <param name="encoder"></param>
+        /// <param name="sharedOptions"></param>
+        /// <param name="options">Configuration options for the middleware.</param>
+        public VkontakteAuthenticationMiddleware(RequestDelegate next, IDataProtectionProvider dataProtectionProvider, ILoggerFactory loggerFactory, UrlEncoder encoder, IOptions<SharedAuthenticationOptions> sharedOptions, IOptions<VkontakteAuthenticationOptions> options)
+            : base(next, dataProtectionProvider, loggerFactory, encoder, sharedOptions, options)
+        {
+            if (next == null)
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationMiddleware.cs]: Remove all these null checks, they are already enforced by the base constructor.

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationOptions.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+            TokenEndpoint = VkontakteAuthenticationDefault.TokenEndpoint;
+            UserInformationEndpoint = VkontakteAuthenticationDefault.GraphApiEndpoint;
+            Fields = new[] { "uid", "first_name", "last_name", "photo_50", "screen_name" };
+        }
+
+        /// <summary>
+        /// The list of fields to retrieve from the UserInformationEndpoint.
+        /// https://vk.com/dev/fields
+        /// </summary>
+        public ICollection<string> Fields { get; }
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationOptions.cs]: `public ICollection<string> Fields { get; } = new HashSet<string>();`

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationOptions.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        /// </summary>
+        public VkontakteAuthenticationOptions()
+        {
+            AuthenticationScheme = VkontakteAuthenticationDefault.AuthenticationScheme;
+            DisplayName = AuthenticationScheme;
+            CallbackPath = new PathString("/signin-vkontakte");
+            AuthorizationEndpoint = VkontakteAuthenticationDefault.AuthorizeEndpoint;
+            TokenEndpoint = VkontakteAuthenticationDefault.TokenEndpoint;
+            UserInformationEndpoint = VkontakteAuthenticationDefault.GraphApiEndpoint;
+            Fields = new[] { "uid", "first_name", "last_name", "photo_50", "screen_name" };
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationOptions.cs]: Use `Add` instead.

File: src/AspNet.Security.OAuth.Vkontakte/project.json [hunk 11]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/project.json]: Note for me: fix this file before merging.

File: src/AspNet.Security.OAuth.Vkontakte/AspNet.Security.OAuth.Vkontakte.xproj [hunk 12]
```diff
@@ -0,0 +1,17 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/AspNet.Security.OAuth.Vkontakte.xproj]: Note for me: fix this file before merging.

**Comment by PinpointTownes** [comment]
Thanks for the PR :tada: 

Overall, it looks fine but please fix the bracing style/XML docs issues.
Feel free to ping me when it's done.

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.Http.Authentication;
+using Microsoft.AspNetCore.WebUtilities;
+using Newtonsoft.Json.Linq;
+
+namespace AspNet.Security.OAuth.Vkontakte
+{
+    internal class VkontakteAuthenticationHandler : OAuthHandler<VkontakteAuthenticationOptions>
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs]: `public`

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHelper.cs [hunk 14]
```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using Newtonsoft.Json.Linq;
+
+namespace AspNet.Security.OAuth.Vkontakte
+{
+    /// <summary>
+    /// Contains static methods that allow to extract user's information from a <see cref="JObject"/>
+    /// instance retrieved from Facebook after a successful authentication process.
+    /// </summary>
+    internal static class VkontakteAuthenticationHelper
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHelper.cs]: `public`

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHelper.cs [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+            {
+                throw new ArgumentNullException(nameof(response));
+            }
+
+            return GetStringValue(response, "photo_50");
+        }
+
+        private static string GetStringValue(JObject jObject, string key)
+        {
+            return jObject["response"][0].Value<string>(key) ?? string.Empty;
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHelper.cs]: We don't want to return `string.Empty`. Remove this method and make all the helpers consistent with the other providers.

**Comment by serber** [comment]
@PinpointTownes, all done

**Comment by PinpointTownes** [comment]
Great, thanks!

Could you also please add the usual LICENSE header to all the .cs files? (e.g https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.BattleNet/BattleNetAuthenticationHandler.cs#L1-L5)

File: src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHelper.cs [hunk 16]
```diff
@@ -0,0 +1,35 @@
+﻿using JetBrains.Annotations;
+using Newtonsoft.Json.Linq;
+
+namespace AspNet.Security.OAuth.Vkontakte {
+    /// <summary>
+    /// Contains static methods that allow to extract user's information from a <see cref="JObject"/>
+    /// instance retrieved from Facebook after a successful authentication process.
+    /// </summary>
+    internal static class VkontakteAuthenticationHelper {
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHelper.cs]: Don't forget to make that public.

**Comment by serber** [comment]
@PinpointTownes, done: headers added, access modifier fixed

**Comment by PinpointTownes** [comment]
Looks fine. I'll merge it tonight and make it's included in the next release.

**Comment by PinpointTownes** [comment]
Merged: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/034dc358a878a17b205758b3cb40475c2434d92b

Thanks for your contribution! :clap:

**Pull request closed without merging**